flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/MutationActionList.java

```diff
@@ -0,0 +1,371 @@
+					}
+					break;
+
+				case INCREMENT:
+					if (increment == null) {
+						increment = new Increment(rowKey);
+						mutations.add(increment);
+					}
+					increment.addColumn(action.getFamily(), action.getQualifier(), action.getIncrement());
+					break;
+
+				case APPEND:
+					if (append == null) {
+						append = new Append(rowKey);
+						mutations.add(append);
+					}
+					append.add(action.getFamily(), action.getQualifier(), action.getValue());
+					break;
+
+				// If there are multiple DELETE_ROW actions, only the first one is served
+				case DELETE_ROW:
+					if (!rowIsDeleted) {
+						for (int i = 0; i < mutations.size(); ) {
+							if (mutations.get(i) instanceof Delete) {
+								mutations.remove(i);
+							} else {
+								i++;
+							}
+						}
+						delete = new Delete(rowKey, action.getTs());
+						mutations.add(delete);
+						rowIsDeleted = true;
+					}
+					break;
+
+				case DELETE_FAMILY:
+					if (!rowIsDeleted) {
+						if (delete == null) {
+							delete = new Delete(rowKey);
+							mutations.add(delete);
+						}
+						delete.addFamily(action.getFamily(), action.getTs());
+					}
+					break;
+
+				case DELETE_COLUMNS:
+					if (!rowIsDeleted) {
+						if (delete == null) {
+							delete = new Delete(rowKey);
+							mutations.add(delete);
+						}
+						delete.addColumns(action.getFamily(), action.getQualifier(), action.getTs());
+					}
+					break;
+
+				case DELETE_COLUMN:
+					if (!rowIsDeleted) {
+						if (delete == null) {
+							delete = new Delete(rowKey);
+							mutations.add(delete);
+						}
+						delete.addColumn(action.getFamily(), action.getQualifier(), action.getTs());
+					}
+					break;
+
+				default:
+					throw new IllegalArgumentException("Cannot process such action type: " + action.getType());
+			}
+		}
+		Durability durability = writeToWAL ? Durability.SYNC_WAL : Durability.SKIP_WAL;
+		for (Mutation mutation : mutations) {
+			mutation.setDurability(durability);
+		}
+		return mutations;
+	}
+
+	/**
+	 * Create a new list of HBase {@link Mutation}s enabling WAL as default.
+	 *
+	 * @param rowKey row that the created {@link Mutation} list is applied to
+	 * @return a list of HBase {@link Mutation}s
+	 */
+	public List<Mutation> newMutationList(byte[] rowKey) {
+		return newMutationList(rowKey, true);
+	}
+
+	/**
+	 * Add to {@link MutationActionList} a {@link MutationAction} of type {@link MutationAction.Type#PUT}. which will
+	 * create an HBase {@link Put} operation for a specified row with specified timestamp.
+	 *
+     * @param family family name
+	 * @param qualifier column qualifier
+	 * @param value column value
+	 * @param timestamp version timestamp
+	 * @return this
+	 */
+	public MutationActionList addPut(byte[] family, byte[] qualifier, byte[] value, long timestamp) {
+		actions.add(new MutationAction(family, qualifier, value, timestamp, 0, MutationAction.Type.PUT));
```

Comment:
The design is to let users define actions they want to perform on a single row given a input record and return a MutationActionList for it. HBaseSink will then call newMutatiionList only once given an input and send associated Mutations to an HBase table. What do you think?